cpp/include/cugraph/algorithms.hpp

```diff
@@ -1125,6 +1125,47 @@ void bfs(raft::handle_t const& handle,
          vertex_t depth_limit      = std::numeric_limits<vertex_t>::max(),
          bool do_expensive_check   = false);
 
+/**
+ * @brief Extract paths from breadth-first search output
+ *
+ * This function extracts paths from the BFS output.  BFS outputs distances
+ * and predecessors.  The path from a vertex v back to the original source vertex
+ * can be extracted by recursively looking up the predecessor vertex until you arrive
+ * back at the original source vertex.
+ * 
+ * If there are multiple source vertices there can be only one per connected component
+ * (as described in the bfs algorithm).
+ *
+ * @throws cugraph::logic_error on erroneous input arguments.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam weight_t Type of edge weights. Needs to be a floating point type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * or multi-GPU (true).
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Graph view object.
+ * @param distances Pointer to the distance array constructed by bfs.
+ * @param predecessors Pointer to the predecessor array constructed by bfs.
+ * @param destinations Destination vertices, extract path from source to each of these destinations
+ * @param n_destinations number of destinations (one source per component at most).
+ * @param paths Pointer to the output path array (preallocated to max_path_length * n_destinations elements).
+ * Paths will be placed into this array in the same order as the destinations input array.  Unused elements
+ * in the paths * will be set to invalid_vertex_id (-1 for a signed vertex_t,
+ * std::numeric_limits<vertex_t>::max() for an unsigned vertex_t type).
+ * @param max_path_length The maximum length of a path to extract.
```

Comment:
What do you think about

taking vertex_t * paths & max_path_length
vs
returning rmm::device_uvector<vertex_t> for paths and rmm::device_uvector<size_t> (or std::vector<size_t>) for offsets?

The latter sounds more functional to me and also users don't need to worry about max_path_length.

Reply:
I definitely like the idea of returning the result rather than making the user pre-allocate it.  That is awkward in the unit test and will be awkward in the python implementation, I suspect.  The implementation can compute max_path_length much more easily.

The current expected uses of this function are python users.  I believe that returning a `num_destinations x max_path_length` dense matrix will be a more appropriate return for these users.

It would probably be useful to either:
* Create an API method to convert a CSR-style result like this into a dense matrix, or
* Create an API method to convert a dense matrix like this into a CSR-style result (or both)

My inclination, given the initial use will be easier with a dense result, would be to return:
`std::tuple<rmm::device_uvector<vertex_t>, vertex_t>`
(the dense matrix of the proper dimensions and the computed value of max_path_length).

Reply:
> My inclination, given the initial use will be easier with a dense result, would be to return:
std::tuple<rmm::device_uvector<vertex_t>, vertex_t>
(the dense matrix of the proper dimensions and the computed value of max_path_length).

I second this, this sounds better than asking users to preallocate.